lib/gauge_hyp.cu

```diff
@@ -0,0 +1,152 @@
+        } else if (level == 3) {
+          launch<HYP>(tp, stream, GaugeHYPArg<Float, nColor, recon, 3, 4>(out, tmp, in, alpha, dir_ignore));
+        }
+      } else if (hypDim == 3) {
+        if (level == 1) {
+          launch<HYP3D>(tp, stream, GaugeHYPArg<Float, nColor, recon, 1, 3>(out, tmp, in, alpha, dir_ignore));
+        } else if (level == 2) {
+          launch<HYP3D>(tp, stream, GaugeHYPArg<Float, nColor, recon, 2, 3>(out, tmp, in, alpha, dir_ignore));
+        }
+      }
+    }
+
+    void preTune() { out.backup(); } // defensive measure in case they alias
+    void postTune() { out.restore(); }
+
+    long long flops() const
+    {
+      long long flops = 0;
+      auto mat_flops = in.Ncolor() * in.Ncolor() * (8ll * in.Ncolor() - 2ll);
+      if ((hypDim == 4 && level == 1) || (hypDim == 3 && level == 1)) {
+        flops += ((hypDim - 1) * 2 + (hypDim - 1) * 4) * mat_flops * hypDim * in.LocalVolume();
+      } else if (hypDim == 4 && level == 2) {
+        flops += ((hypDim - 1) * 2 + (hypDim - 1) * (hypDim - 2) * 4) * mat_flops * hypDim * in.LocalVolume();
+      } else if ((hypDim == 4 && level == 3) || (hypDim == 3 && level == 2)) {
+        flops += (2 + (hypDim - 1) * 4) * mat_flops * hypDim * in.LocalVolume();
+      }
+      return flops;
+    }
+
+    long long bytes() const
+    {
+      long long bytes = 0;
+      if ((hypDim == 4 && level == 1) || (hypDim == 3 && level == 1)) { // 6 links per dim, 1 in, hypDim-1 tmp
+        bytes += (in.Reconstruct() * in.Precision() + (hypDim - 1) * 6 * in.Reconstruct() * in.Precision()
+                  + (hypDim - 1) * tmp[0]->Reconstruct() * tmp[0]->Precision())
+          * hypDim * in.LocalVolume();
+      } else if (hypDim == 4 && level == 2) { // 6 links per dim, 1 in, hypDim-1 tmp
+        bytes += (in.Reconstruct() * in.Precision()
+                  + (hypDim - 1) * (hypDim - 2) * 6 * tmp[0]->Reconstruct() * tmp[0]->Precision()
+                  + (hypDim - 1) * tmp[0]->Reconstruct() * tmp[0]->Precision())
+          * hypDim * in.LocalVolume();
+      } else if ((hypDim == 4 && level == 3) || (hypDim == 3 && level == 2)) { // 6 links per dim, 1 in, 1 out
+        bytes += (in.Reconstruct() * in.Precision() + (hypDim - 1) * 6 * tmp[0]->Reconstruct() * tmp[0]->Precision()
+                  + out.Reconstruct() * out.Precision())
+          * hypDim * in.LocalVolume();
+      }
+      return bytes;
+    }
+
+  }; // GaugeAPE
+
+  void HYPStep(GaugeField &out, GaugeField &in, double alpha1, double alpha2, double alpha3, int dir_ignore)
+  {
+    checkPrecision(out, in);
+    checkReconstruct(out, in);
+    checkNative(out, in);
+
+    GaugeFieldParam gParam(out);
+    gParam.location = QUDA_CUDA_FIELD_LOCATION;
+    const int smearDim = (dir_ignore >= 0 && dir_ignore <= 3) ? 3 : 4;
+    // GaugeField tmp[4];
+    gParam.geometry = QUDA_TENSOR_GEOMETRY;
+    GaugeFieldParam gParam2(gParam);
```

Comment:
@SaltyChiang @weinbe2 this is an unused variable and should be deleted: it's the cause of warnings causing some of the CI targets to fail.